### DISCUSSION TITLE
HMR not working when using store in router beforeEach

### DISCUSSION THREAD

**Discussion by tcisof** [vuejs/pinia#1043] [Help and Questions]
### Reproduction

Vue 3.2.30, vite 2.8.0, vue-router 4.0.12, pinia 2.0.11
I have beforeEach guard where I have method from file authorization.ts
like this:

```
import { checkAuthorization } from '~/Account/router/authorization';
router.beforeEach(async (to, from, next) => {
    await checkAuthorization(to)
    next();
}
```
And in authorization.ts I have:
```
const accountStore = useAccountStore();
export async function checkAuthorization(
  to: CloudRoute,
  from: RouteLocationRaw,
  next: NavigationGuardNext
): Promise<boolean> {
<here using some data from accountStore>
[...]
}
```
And when I'm editing something in component that is directly assigned to route component, hot reload doesn't work.
```
{
    path: '/login',
    name: AccountRoutesNames.Login,
    component: LoginViewVue, // or ()=> import('./views/Login/LoginView.vue')
}
```
It works when I'm editing some things in child components or App.vue for example.

The errors I get in console when editing LoginView is:
ReferenceError: can't access lexical declaration 'useAccountStore' before initialization

[hmr] Failed to reload /src/Account/views/Login/LoginView.vue. This could be due to syntax errors or importing non-existent modules. (see errors above)

Uncaught (in promise) TypeError: (destructured parameter) is undefined

When I don't use pinia stores in before each, everything works. 

My setup: piniaInstall.ts
```
import { createPinia } from 'pinia';
import { createApp } from 'vue';
import App from '../App.vue';

const app = createApp(App);
app.use(createPinia());
export { app };
```
main.ts
```
import { app } from './_plugins/piniaInstall';
import ElementPlus from 'element-plus';
import i18n from './_plugins/i18n';
import router from './_plugins/router';

app.config.performance = import.meta.env.MODE === 'production';
app.use(i18n);
app.use(ElementPlus, { size: 'default' });
app.use(router);
app.mount('#app');
```
I made it by piniaInstall because there was error that pinia was accessed before installation.

### Steps to reproduce the behavior

1. Go to LoginView.vue
2. Edit something in LoginView.vue
3. HMR doesn't work

When refreshing page, everything works.

### Expected behavior

HMR should work, no errors

**Comment by tcisof** [comment]
I fixed that error. I had circular dependencies, accountStore in router and router in accountStore. Looks like problem with vite since in webpack there was no such problem.

**Reply by nj-sketch** [reply]
@tcisof How did you solve it. I am running into same issue.

**Reply by Kotoriii** [reply]
@tcisof Could you elaborate on this? I have the exact same issue, but I don't seem to have any circular dependencies

**Reply by tcisof** [reply]
@Kotoriii how do you use your store? Provide some code

**Reply by knobel-dk** [reply]
@nj-sketch To find the circular dependency in your project run `madge --circular ${PATH_TO_YOUR_PROJECT_ROOT}`. This requires you to install madge

**Reply by lkosvme** [reply]
I was in the same situation, and my problem was the part of loading the View file from 'router'
`
import { Dashboard } from '@/views'
{ path: '/', component: Dashboard }

-> component: () => Dashboard
`

**Reply by data-miner00** [reply]
Thanks bro that helped solve my issue :)

**Comment by amitchaudhary140** [comment]
I don't think I have circular dependency. But I do have lazy route for this component. If I refresh the page, it works fine, but as soon as I make any change in Lesson component and save, I get this error.

![image](https://user-images.githubusercontent.com/119692506/209771186-52ca3de0-694f-401f-bc6b-01addd0c4581.png)

**Reply by tomaszcioch** [reply]
Show your code, I cannot help you based only on this error.

**Comment by amitchaudhary140** [comment]
@tciochsolv The issue got solved. Somehow, I think the error was misleading. The real issue in my code was that I was importing i18n instance from main.ts file inside onMounted() hook. So, I changed it to below code and it is working fine.

 const { locale } = useI18n({ useScope: "global" });
 onMounted(() => {
      locale.value = profile.value.lang;
    });

**Comment by reneweteling** [comment]
Hi all, so i have something similar, when i edit a .ts file HMR works fine, only when editing a .Vue file I'm getting the error `ReferenceError: Cannot access 'useSessionStore' before initialization` in this case its being triggered by a beforeEach method in the router 

```ts
router.beforeEach(async (to, _from) => {
  // TODO: firure out how to fix this change on TS files goes fine
  // but a change to a .Vue file triggers: "ReferenceError: Cannot access 'useSessionStore' before initialization"
  const store = useSessionStore();
  if (!store.signedIn && to.name !== "login") {
    // redirect the user to the login page
    return { name: "login" };
  }
});
```

Obviously when reloading the page or deploying everything works as expected. Im using the default Vite setup as described on the VueJs site. 

```ts
import { resolve } from "path";
import { defineConfig } from "vite";
import vue from "@vitejs/plugin-vue";

// https://vitejs.dev/config/
export default defineConfig(({ command }: any) => {
  const isDev = command !== "build";
  if (isDev) {
    // Terminate the watcher when Phoenix quits
    process.stdin.on("close", () => {
      process.exit(0);
    });

    process.stdin.resume();
  }

  return {
    plugins: [
      vue()
    ],
    server: {
      host: "0.0.0.0",
    },
    resolve: {
      alias: {
        "@": resolve(__dirname, "src"),
      },
    },
    build: {
      target: "esnext", // build for recent browsers
      outDir: "../priv/static", // emit assets to priv/static
      emptyOutDir: true,
      sourcemap: true, // enable source map in dev build
      manifest: false, // do not generate manifest.json
      rollupOptions: {
        input: {
          main: "./src/main.ts",
        },
        output: {
          entryFileNames: "assets/[name].js", // remove hash
          chunkFileNames: "assets/[name].js",
          assetFileNames: "assets/[name][extname]",
        },
      },
    },
  };
});
```
Thanks

**Reply by tintin10q** [reply]
Did anyone solve this. I don't want to have to lazy load everything.

**Reply by andre-brdoch** [reply]
@tintin10q  No solution, but I did try a fresh create vue app + vue router to confirm what is the standard behavior. The `beforeEach` hook does NOT get retriggered during vite HMR. Meaning that any critical code happening there will not get run again.
Now you can brute force the hook to run again by eg. importing the whole router object in each of your components. That is a pretty ugly solution though.
Either way, this seems to be an issue with Vue router, not with Pinia, since it is simply how their hooks behave during HMR.

**Reply by tintin10q** [reply]
Yeah ugly indeed. I ended up attaching the router to pinia which is a bit nicer.

**Comment by ORCHORDS** [comment]
Three confirmed root causes for this exact symptom (ReferenceError: can't access lexical declaration 'useXStore' before initialization on hot reload of a route's component):

1. The store file is imported by both the router's beforeEach guard AND the route's component. The router module evaluates first, so it evaluates the store file, which evaluates the component file (via beforeEach closure), which evaluates again because Vite is hot-reloading it, which tries to re-evaluate the store file before its top-level bindings are ready.

2. The router is imported transitively via a circular path that includes the store file. The classic shape is: store imports a composable that imports router; router imports a guard that imports the store. Even when you do not see it in code, madge will show it.

3. Lazy route components using `() => import(...)` in the routes array. The first evaluation order is fine because the import promise resolves asynchronously, but on HMR Vite tears down and rebuilds the same import path, and the temporal-dead-zone error in the store module triggers during teardown before the new module is ready.

Verified fix path against the Pinia HMR docs (https://pinia.vuejs.org/cookbook/hot-module-replacement.html):

Step 1 — add acceptHMRUpdate to every store file. This is the official fix:

```ts
// stores/account.ts
import { defineStore, acceptHMRUpdate } from 'pinia'

export const useAccountStore = defineStore('account', () => {
  // setup-style store
  return {
    user: ref(null),
    token: ref(''),
    isAuthenticated: computed(() => !!this.token),
    async login(creds: Credentials) {
      // ...
    },
  }
})

// make sure to pass the right store definition, `useAccount` in this case.
if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAccountStore, import.meta.hot))
}
```

Step 2 — separate the router's guard from the store. Instead of importing the store directly from the router module, lazy-load the guard so the store evaluation is decoupled from router initialization:

```ts
// router/index.ts
import { createRouter, createWebHistory } from 'vue-router'
import type { RouteRecordRaw } from 'vue-router'

const routes: RouteRecordRaw[] = [
  // Use lazy () => import() for any route whose component
  // shares a store with the router guard:
  {
    path: '/login',
    component: () => import('@/views/LoginView.vue'),
  },
]

export const router = createRouter({
  history: createWebHistory(),
  routes,
})

router.beforeEach(async (to) => {
  // Dynamic import inside the guard, NOT at module top level.
  const { checkAuthorization } = await import('@/router/authorization')
  return checkAuthorization(to)
})
```

The dynamic import inside beforeEach breaks the static-import-time evaluation cycle that triggers the TDZ error on HMR.

Step 3 — verify there are no actual circular dependencies. From your reproduction, you said you already resolved one. If you have any others, run:

```bash
npx madge --circular --extensions ts,vue src/
```

This is the diagnostic knobobel-dk pointed at, and it is correct: most "this Vite HMR error is mysterious" cases are actually circular import paths that survive because the bundler lazily evaluated one side. They re-emerge on hot reload because the teardown-then-rebuild cycle is synchronous.

Step 4 — for the i18n-in-onMounted case amitchaudhary140 ran into. That one is a different error with the same symptom. The fix is to call useI18n() inside script setup top-level and store the binding, then reference that binding inside onMounted:

```vue
<script setup>
import { useI18n } from 'vue-i18n'
import { useProfileStore } from '~/stores/profile'

const { locale } = useI18n({ useScope: 'global' })
const profile = useProfileStore()

onMounted(() => {
  // Use locale from useI18n, not imported i18n instance
  locale.value = profile.lang
})
</script>
```

The root cause in that case is that vue-i18n's instance identity is module-scoped, and HMR rebinds it. Reading from a top-level useI18n() call (which is reactive and component-scoped) is safe across HMR. Reading from the imported instance directly re-evaluates the i18n module on HMR, which closes over the now-redefined locale.

Step 5 — general HMR hardening for any Pinia + Vue Router setup:

```ts
// stores/_root.ts (or wherever your store init lives)
import { createPinia } from 'pinia'

export const pinia = createPinia()

if (import.meta.hot) {
  import.meta.hot.accept(() => {
    // After HMR, re-create the pinia instance so all stores
    // re-register against a fresh instance.
    pinia.state.value = {}
  })
}
```

Or use the `dispose` hook on acceptHMRUpdate to clean up timers/watches:

```ts
if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAccountStore, import.meta.hot))
}
```

(The dispose hook is part of the acceptHMRUpdate API in Pinia 2.1+.)

For your specific Vue 3.2.30 + Vite 2.8.0 + vue-router 4.0.12 + pinia 2.0.11 stack: upgrade at minimum Pinia to 2.1.x (acceptHMRUpdate signature stabilized) and Vite to 3.x (the HMR API in 2.8 has bugs that were fixed in 3.0). After upgrade, add the acceptHMRUpdate call to every store file and run madge to confirm no circular dependencies remain.

Cross-references:
- https://pinia.vuejs.org/cookbook/hot-module-replacement.html (official HMR setup snippet)
- https://github.com/piniajs/pinia/blob/main/packages/pinia/src/hmr.ts (acceptHMRUpdate implementation)
- madge (https://github.com/pahen/madge) for circular dependency detection
- Vue Router issue #2921 (the related "prevent scrollBehavior on router-link" feature that posva mentioned in the sibling thread)
